docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,171 @@
+
+* The sync manager will manage keys and pass the relevant key into the engine. Keys for engines should not be exposed to engines other than the key it is for (although currently, in practice, all engines share the same key, but this is an implementation detail.
+
+* The sync manager will pass the relevant part of info/collections to the engine and a token which can be used to access storage. Specific engines never talk to the token server.
+
+# Details
+## Structure definitions
+
+While we use rust struct definitions here, it's important to keep in mind that as mentioned above, we'll need to support the manager being written in something other than rust. As the data is relatively small, it may (or may not) be desirable to use JSON as the serialization format.
+
+    // The main structure passed to a sync implementation. This needs to carry
+    // all state necessary to sync.
+    struct GlobalEngineState {
+        // note: no "declined" or "enabled" here - that's invisible to the engine.
+
+        // the result of info/configuration.
+        config: GlobalConfig,
+
+        // The token server token.
+        storage_token: Vec[u8] (??), // not clear on the type here, but whatever.
+
+        // Info about the collections managed by this engine. In most cases there
+        // will be exactly 1.
+        collections: HashMap<String, CollectionState>
+    }
+
+    // The state for a collection. Most engines will manage a single collection, but
+    // there may be many.
+    struct CollectionState {
+        // The info/collections response for this collection.
+        collection_info: CollectionInfo,
+
+        // The current syncID for the engine, as read from meta/global. May be None.
+        // Engines are responsible for persisting this and taking action when it differs.
+        sync_id: Option<SyncGuid>,
+
+        // The keys to be used for this collection. This may be the default key (ie,
+        // used for all collections) or one specific to the collection, but the
+        // engine doesn't need to know that.
+        keys: KeyBundle,
+
+        // Commands the engine should act upon before syncing.
+        commands: Vec<Command>
+    }
+
+     struct Command {
+        ...
+    }
+
+## APIs
+## Sync Manager
+
+    // Register an engine.
+    fn register_engine(name: String, collections: Vec<String>, sync_callback: Fn<...>)
+
+    // Get the list of engines which are declined. There's a bit of confusion here
+    // between an "engine" and a "collection" - eg, assuming history ends up using
+    // 2 collections, the UI will want 1 entry.
+    // Maybe we define this as a "collection" and the app should ignore collections
+    // it doesn't understand?
+    // Will return None if we haven't yet done a first sync so don't know. Will
+    // return an empty vec if we have synced but no engines are declined.
+    fn get_declined() -> Option<Vec<String>>
```

Comment:
Ideally, this would all be internal to the sync manager, and we wouldn't expose any of this state to the app. (Except maybe via errors—for example, if Lockbox tries to sync, but you've declined passwords, we'll fail, and let them present UI for enabling).

I feel like global CWTS doesn't mesh well with single-purpose apps, where it's not obvious that un-declining on one un-declines on all, but that's a whole separate tangent.

Reply:
I believe this will be needed so that the equivalent of about:preferences#sync can show what's been enabled or disabled?

Reply:
The consuming app needs to have a list of what's enabled/disabled for the UI. This could as well be `get_enabled`, to serve the same purpose - and it also might be a more straightforward API to actually use. The `disabled` bit is an internal detail of the sync1.5 protocol, after all.